framework/decode/vulkan_address_replacer.cpp

```diff
@@ -0,0 +1,791 @@
+** FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+** DEALINGS IN THE SOFTWARE.
+*/
+
+#include "decode/vulkan_address_replacer.h"
+#include "decode/vulkan_address_replacer_shaders.h"
+#include "decode/mark_injected_commands.h"
+#include "util/logging.h"
+
+GFXRECON_BEGIN_NAMESPACE(gfxrecon)
+GFXRECON_BEGIN_NAMESPACE(decode)
+
+//! RAII helper to mark injected commands in scope
+struct mark_injected_commands_helper_t
+{
+    mark_injected_commands_helper_t()
+    {
+        // mark injected commands
+        decode::BeginInjectedCommands();
+    }
+
+    ~mark_injected_commands_helper_t()
+    {
+        // mark end of injected commands
+        decode::EndInjectedCommands();
+    }
+};
+
+inline uint32_t aligned_size(uint32_t size, uint32_t alignment)
+{
+    return (size + alignment - 1) & ~(alignment - 1);
+}
+
+inline uint32_t div_up(uint32_t nom, uint32_t denom)
+{
+    GFXRECON_ASSERT(denom > 0)
+    return (nom + denom - 1) / denom;
+}
+
+uint32_t get_memory_type_index(const VkPhysicalDeviceMemoryProperties& memory_properties,
+                               uint32_t                                type_bits,
+                               VkMemoryPropertyFlags                   property_flags)
+{
+    uint32_t memory_type_index = std::numeric_limits<uint32_t>::max();
+
+    for (uint32_t i = 0; i < memory_properties.memoryTypeCount; ++i)
+    {
+        if ((type_bits & (1 << i)) &&
+            ((memory_properties.memoryTypes[i].propertyFlags & property_flags) == property_flags))
+        {
+            memory_type_index = i;
+            break;
+        }
+    }
+
+    return memory_type_index;
+}
+
+//! 16 bytes
+struct hashmap_t
+{
+    VkDeviceAddress storage;
+    uint32_t        size;
+    uint32_t        capacity;
+};
+
+struct replacer_params_t
+{
+    hashmap_t hashmap;
+
+    // input-/output-arrays can be identical when sbt-alignments/strides match
+    VkDeviceAddress input_handles, output_handles;
+    uint32_t        num_handles;
+};
+
+decode::VulkanAddressReplacer::buffer_context_t::~buffer_context_t()
+{
+    if (resource_allocator != nullptr)
+    {
+        if (buffer != VK_NULL_HANDLE)
+        {
+            resource_allocator->DestroyBufferDirect(buffer, nullptr, allocator_data);
+        }
+        if (device_memory != VK_NULL_HANDLE)
+        {
+            resource_allocator->FreeMemoryDirect(device_memory, nullptr, memory_data);
+        }
+    }
+}
+
+VulkanAddressReplacer::VulkanAddressReplacer(const VulkanDeviceInfo*              device_info,
+                                             const encode::VulkanDeviceTable*     device_table,
+                                             const decode::CommonObjectInfoTable& object_table) :
+    device_table_(device_table)
+{
+    GFXRECON_ASSERT(device_info != nullptr && device_table != nullptr)
+
+    const VulkanPhysicalDeviceInfo* physical_device_info = object_table.GetVkPhysicalDeviceInfo(device_info->parent_id);
```

Comment:
Should we do a `mark_injected_commands_helper_t` here as well?

Reply:
we could. I kept it without the helper here, because I liked the readability better. it was more what's injected (mostly because the scope is rather long)